module/mempool/stdmap/backdata/arraylinkedlist/entitiesLinkedList.go

```diff
@@ -0,0 +1,350 @@
+// [0, limit). In other words, all slice indices in cachedEntity are marked as free indices.
+func (e *EntityDoubleLinkedList) initFreeEntities(limit uint32) {
+	e.free.head.setPointer(0)
+	e.free.tail.setPointer(0)
+
+	for i := uint32(1); i < limit; i++ {
+		// appends slice index i to tail of free linked list
+		e.link(e.free.tail, EIndex(i))
+		// and updates its tail
+		e.free.tail.setPointer(EIndex(i))
+	}
+}
+
+// Add writes given entity into a cachedEntity on the underlying entities linked-list. Return value is
+// the index at which given entity is written on entities linked-list so that it can be accessed directly later.
+func (e *EntityDoubleLinkedList) Add(entityId flow.Identifier, entity flow.Entity, owner uint64) EIndex {
+	entityIndex := e.sliceIndexForEntity()
+	e.values[entityIndex].entity = entity
+	e.values[entityIndex].id = entityId
+	e.values[entityIndex].owner = owner
+	e.values[entityIndex].node.next.setUndefined()
+	e.values[entityIndex].node.prev.setUndefined()
+
+	if e.used.head.isUndefined() {
+		// used list is empty, hence setting head of used list to current entityIndex.
+		e.used.head.setPointer(entityIndex)
+		e.values[e.used.head.sliceIndex()].node.prev.setUndefined()
+	}
+
+	if !e.used.tail.isUndefined() {
+		// links new entity to the tail
+		e.link(e.used.tail, entityIndex)
+	}
+
+	// since we are appending to the used list, entityIndex also acts as tail of the list.
+	e.used.tail.setPointer(entityIndex)
+
+	e.total++
+	return entityIndex
+}
+
+// Get returns entity corresponding to the entity index from the underlying list.
+func (e EntityDoubleLinkedList) Get(entityIndex EIndex) (flow.Identifier, flow.Entity, uint64) {
+	return e.values[entityIndex].id, e.values[entityIndex].entity, e.values[entityIndex].owner
+}
+
+// sliceIndexForEntity returns a slice index which hosts the next entity to be added to the list.
+func (e *EntityDoubleLinkedList) sliceIndexForEntity() EIndex {
+	if e.free.head.isUndefined() {
+		// free list is empty, hence, we are out of space, we
+		// need to eject.
+		if e.ejectionMode == RandomEjection {
+			// turning a random entity into a free head.
+			e.invalidateRandomEntity()
+		} else {
+			// LRU ejection
+			// used head is the oldest entity, hence,
+			// turning the used head to a free head.
+			e.invalidateUsedHead()
+		}
+	}
+
+	// claiming the head of free list as the slice index for the next entity to be added
+	return e.claimFreeHead()
+}
+
+// Size returns total number of entities that this list maintains.
+func (e EntityDoubleLinkedList) Size() uint32 {
+	return e.total
+}
+
+// getHeads returns entities corresponding to the used and free heads.
+func (e EntityDoubleLinkedList) getHeads() (*cachedEntity, *cachedEntity) {
+	var usedHead, freeHead *cachedEntity
+	if !e.used.head.isUndefined() {
+		usedHead = &e.values[e.used.head.sliceIndex()]
+	}
+
+	if !e.free.head.isUndefined() {
+		freeHead = &e.values[e.free.head.sliceIndex()]
+	}
+
+	return usedHead, freeHead
+}
```

Comment:
I'm not sure why it would ever be necessary to return both heads or both tails. I've worked on code like this before and this function is not something that has been necessary in the past. Maybe this is a sign suggesting that the code could profit from abstracting to support an arbitrary number of states as suggested here [1]?

And even if it is necessary, then abstracting `free` and `used` into an array of states would de-duplicate the inner workings of these functions, or?

[1] https://www.notion.so/dapperlabs/Intro-to-heap-friendly-hero-structures-cd613aba6f4244f88e58fec1f222f476

Reply:
We typically use this function to retrieve `usedHead`. The returned `freeHead` is used only for testing. Anyway, the method is internal to the package and is not exported. Also, as mentioned earlier, issued up the state-based abstraction here to address separately: dapperlabs/flow-go#6083